src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -4001,7 +3996,16 @@ public static void ComputeJitPgoInstrumentationSchema(Func<object, IntPtr> objec
 
             if (!_pgoResults.TryGetValue(methodDesc, out PgoInstrumentationResults pgoResults))
             {
-                var pgoResultsSchemas = getPgoInstrumentationResults(methodDesc);
+#if READYTORUN
+                PgoSchemaElem[] pgoResultsSchemas = _compilation.ProfileData.GetAllowSynthesis(_compilation, methodDesc, out bool isSynthesized)?.SchemaData;
+
+                if (pgoResultsSchemas != null && isSynthesized && _compilation.ProfileData.EmbedPgoDataInR2RImage)
+                {
+                    _methodCodeNode.AddSynthesizedPgoDataDependency(pgoResultsSchemas);
```

Comment:
@davidwrighton The way this currently works is that the `GetAllowSynthesis` call above will both synthesize the schema and "register" it to be physically embedded once `InstrumentationDataTableNode.GetData` is called. This `AddSynthesizedPgoDataDependency` will then cause `MethodWithGCInfo` to return the schema's dependencies from `MethodWithGCInfo.ComputeNonRelocationBasedDependencies`.

It seems to work, but I wonder if the approach is problematic in case the compilation here fails. Do I need to instead register the schema for physical embedding from inside `MethodWithGCInfo.PublishCode` to make sure the physical embedding and `MethodWithGCInfo.ComputeNonRelocationBasedDependencies` are kept "in sync"?

Reply:
We may run compilation for a single method more than once. Thus the best practice is to not affect any other part of the system untile the PublishCode logic is run.